### DISCUSSION TITLE
Verfi JWT from ETHR fails

### DISCUSSION THREAD

**Discussion by RobertoSnap** [decentralized-identity/veramo#579] [Q&A] [answered]
Creating a VC with createVerifiableCredential()
`bash
 {
      credentialSubject: {
        name: 'Ola Nordman',
        id: 'did:ethr:brok:0x03719c5f561b5342216fd3b204d890cf157f192f7bf40ed3f9301c5ca05690726d'
      },
      issuer: {
        id: 'did:ethr:brok:0x02bde9847ad8569df159b5783df389c2db4d00c25a4c4822eea6fb9963b096e16d'
      },
      type: [ 'VerifiableCredential' ],
      '@context': [ 'https://www.w3.org/2018/credentials/v1' ],
      issuanceDate: '2021-06-18T08:06:18.000Z',
      proof: {
        type: 'JwtProof2020',
        jwt: 'eyJhbGciOiJFUzI1NksiLCJ0eXAiOiJKV1QifQ.eyJ2YyI6eyJjcmVkZW50aWFsU3ViamVjdCI6eyJuYW1lIjoiT2xhIE5vcmRtYW4ifSwiQGNvbnRleHQiOlsiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvdjEiXSwidHlwZSI6WyJWZXJpZmlhYmxlQ3JlZGVudGlhbCJdfSwic3ViIjoiZGlkOmV0aHI6YnJvazoweDAzNzE5YzVmNTYxYjUzNDIyMTZmZDNiMjA0ZDg5MGNmMTU3ZjE5MmY3YmY0MGVkM2Y5MzAxYzVjYTA1NjkwNzI2ZCIsIm5iZiI6MTYyNDAwMzU3OCwiaXNzIjoiZGlkOmV0aHI6YnJvazoweDAyYmRlOTg0N2FkODU2OWRmMTU5YjU3ODNkZjM4OWMyZGI0ZDAwYzI1YTRjNDgyMmVlYTZmYjk5NjNiMDk2ZTE2ZCJ9.UYVR5h5dSv1QUqZM3GsQvfVlGlkJpFf4c2_mREIWxG7PL9-AXw7nJhQRHYr7mwa2BqRDNEjxNBnX4SyUG6GIYg'
      }
    }
`

Then checking the JWT inside VC with

```ts
 try {
      await this.agent.handleMessage({
        raw: jwt,
      });
      return true;
    } catch (error) {
      console.log('JWT not valid => ', error);
      return false;
    }
```

Getting error:
```bash
JWT not valid =>  Error: Unsupported message type
        at Object.<anonymous> (/Users/robin/code/brok-auth-bot/node_modules/@veramo/message-handler/src/abstract-message-handler.ts:4:44)
        at Runtime._execModule (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:1299:24)
        at Runtime._loadModule (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:898:12)
        at Runtime.requireModule (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:746:10)
        at Runtime.requireModuleOrMock (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:919:21)
        at Object.<anonymous> (/Users/robin/code/brok-auth-bot/node_modules/@veramo/message-handler/src/index.ts:8:1)
        at Runtime._execModule (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:1299:24)
        at Runtime._loadModule (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:898:12)
        at Runtime.requireModule (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:746:10)
        at Runtime.requireModuleOrMock (/Users/robin/code/brok-auth-bot/node_modules/jest-runtime/build/index.js:919:21)
````

Here is the test that fails:

https://github.com/BROKLab/auth-bot/blob/veramo/src/network/veramo.service.spec.ts#L33

Same thing worked when doing it with a KEY identifier.

Does this handle otherwise with ETHR?

**Comment by mirceanis** [comment, marked as answer]
After some digging it seems there is an issue with the JsonRPC provider. It's not able to confirm the network ID.
I will dig a little deeper, since this might be an issue with `ethr-did-resolver`.

Using your code with a known JsonRPC (like one from infura) seems to work ok.

Update:
 - I'm getting a HTTP 401 error when using the JsonRPC url you provide in your tests project to call 'eth_chainId' or 'net_version'.

**Reply by RobertoSnap** [reply]
Yea, the RPC endpoint has a http basic auth attached, so if the http client does not base64(username, password) and add it as header it will get 401, thats probably the issue. It responds to both "eth_chainId" and "net_version" when these are added. 

I know at least Ethers does this for http request, but must check what http client the ethr-did-resolver uses.

**Reply by RobertoSnap** [reply]
Looks like it uses Ethers: https://github.com/decentralized-identity/ethr-did-resolver/blob/9372e486920cf44f1ed20beff82d3687255d380e/src/configuration.ts#L56

🤔

**Reply by mirceanis** [reply]
yeah, it uses the `JsonRpcProvider` from ethers when supplied with `rpcUrl`. I'm not sure yet how it can be configured to attach the header you mention, but if you know already, please suggest it.

Also, the network configuration for both `ethr-did-resolver` and the `@veramo/did-provider-ethr` should accept a preconfigured `provider` or `web3Provider` property instead of `rpcUrl`.

**Reply by RobertoSnap** [reply]
Ahh, yea, its must be explicit:

```ts
new ethers.providers.JsonRpcProvider({
              url: 'https://e0mvr9jrs7-e0iwsftiw5-rpc.de0-aws.kaleido.io',
              user: 'e0avzugh9j',
              password: '5VOuyz9VPLenxC-zB2nvrWOlfDrRlSlcg0VZyIAvEeI',
            })
```

Will see if i can pass i through

**Reply by RobertoSnap** [reply]
Yea, worked https://github.com/BROKLab/auth-bot/blob/veramo/src/network/veramo.utils.ts#L59

Thanks!

**Reply by vijay05155** [reply]
hi everyone, please help me on this issue,  i was hosted a server to handle the messages but when some agent send a messages to my server agent , then my server agent crash with this error ; "Unsupported message type"
then i was tried to handle the message in my server like : 
		
		 const raw = "eyJhbGciOiJFUzI1NksiLCJ0eXAiOiJKV1QifQ.eyJ2YyI6eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiUHJvZmlsZSJdLCJjcmVkZW50aWFsU3ViamVjdCI6eyJuYW1lIjoidmlqYXkiLCJhZ2UiOiIxOCJ9fSwic3ViIjoiZGlkOmV0aHI6cmlua2VieToweDAzZTc4NmMwYTQzMDdiNmYzYzEzNzk1OThmYTY3OTc3ODVlNTRjODM4MjI2Yzk4MTgwYzFkMjc5MjJlMTMzMGJiMiIsIm5iZiI6MTY1MzI4NDcyMSwiaXNzIjoiZGlkOmV0aHI6cmlua2VieToweDAyYjk5YTEyN2M5ZDhkZTUzNWY5ZWExNWNhODcyNjEyZmRjMWNmODVjM2U3MzVjYTc5NTY1NTc0MDAzM2I0OTZlOCJ9.FBws7eGC_qyjSudD1vMsWriBqyqtGtgZKzoPXUoZconos7Fv9_Zt3RDGdKwRHxM4-dt4olRXQYMxY-K48AV-Ag"
			const parsedMessage = await agent.handleMessage({
			raw,
			save: false,
			metaData: [{ type: 'test' }],
		})
			console.log(parsedMessage);
	
		it' successfully return the VC so what was the issue?

**Reply by vadimchilinciuc** [reply]
you fixed it ?
